lib/Dialect/Seq/SeqPasses.cpp

```diff
@@ -39,38 +40,148 @@ struct CompRegLower : public OpConversionPattern<CompRegOp> {
 
   LogicalResult
   matchAndRewrite(CompRegOp reg, ArrayRef<Value> operands,
-                  ConversionPatternRewriter &rewriter) const final {
-    Location loc = reg.getLoc();
-
-    auto svReg = rewriter.create<sv::RegOp>(loc, reg.getResult().getType());
-    DictionaryAttr regAttrs = reg->getAttrDictionary();
-    if (!regAttrs.empty())
-      svReg->setAttrs(regAttrs);
-    if (!svReg->hasAttrOfType<StringAttr>("name"))
-      // sv.reg requires a name attribute.
-      svReg->setAttr("name", rewriter.getStringAttr(""));
-    auto regVal = rewriter.create<sv::ReadInOutOp>(loc, svReg);
-
-    if (reg.reset() && reg.resetValue()) {
-      rewriter.create<sv::AlwaysFFOp>(
-          loc, sv::EventControl::AtPosEdge, reg.clk(), ResetType::SyncReset,
-          sv::EventControl::AtPosEdge, reg.reset(),
-          [&]() { rewriter.create<sv::PAssignOp>(loc, svReg, reg.input()); },
-          [&]() {
-            rewriter.create<sv::PAssignOp>(loc, svReg, reg.resetValue());
-          });
-    } else {
-      rewriter.create<sv::AlwaysFFOp>(
-          loc, sv::EventControl::AtPosEdge, reg.clk(),
-          [&]() { rewriter.create<sv::PAssignOp>(loc, svReg, reg.input()); });
+                  ConversionPatternRewriter &rewriter) const final;
+};
+
+} // anonymous namespace
+
+namespace {
+class RegLower : public OpConversionPattern<RegOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(RegOp reg, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final;
+};
+
+} // anonymous namespace
+
+struct RegOperandsAndAttrs {
```

Comment:
Could you define and use a custom op interface instead of this? Looks like that's essentially what you're doing.